### PR TITLE
Support for CUDA 9.0

### PR DESCRIPTION
**Pull request by nluehr**
For review by @zheng-xq.

Add explicit __syncwarp to bias_op
 - Makes warp-synchronous code safe on Volta
Add sync mask to __shfl intrinsics
Add libdevice bytecode paths for CUDA 9
 - In CUDA 9, all supported architectures are merged into a single file
Update code gating for CUDA 9
Add sm_70 to the lookup table used by XLA
Change the default sm arch from 20 to 30.
Fix for NVPTX not yet supporting sm_70
Remove unnecessary cuda decorators from defaulted constructors
Use updated NCCL for CUDA 9 fp16 support

### CODE REVIEW

**Comment by jhseu** [comment]
Please handle the CLA before we look at this change. Thanks!

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc [hunk 1]
```diff
@@ -1153,8 +1154,13 @@ __launch_bounds__(1024, 2) void DepthwiseConv2dBackpropFilterGPUKernelNHWCSmall(
       if (filter_depth < in_depth) {
         T val = accum_data[i];
         // Warp-accumulate the pixels of the same depth from the accumulator.
+        int lane_id;
+        asm volatile ("mov.u32 %0, %laneid;" : "=r"(lane_id));
```
Comment (jlebar) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: In the past we've used the threadIdx % 32.  Does that no longer work?

It's kind of a layering violation to put inline assembly here.  If there's no good way to get the lane ID other than inline asm, it should be abstracted away in a function.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc [hunk 2]
```diff
@@ -1116,6 +1116,7 @@ __launch_bounds__(1024, 2) void DepthwiseConv2dBackpropFilterGPUKernelNHWCSmall(
 
     // Note: the condition to reach this is uniform across the entire block.
     __syncthreads();
+    unsigned active_threads = CudaBallot(0xFFFFFFFF, depth_in_range);
```
Comment (jlebar) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: Can we make 0xFFFFFFFF a constant?  It seems pretty bad if someone were to leave off one of the F's.  Same elsewhere.

Or perhaps the order of args to CudaBallot should be reversed, and 0xFFFFFFFF should be the default for the (now) second argument?

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc [hunk 3]
```diff
@@ -1153,8 +1154,13 @@ __launch_bounds__(1024, 2) void DepthwiseConv2dBackpropFilterGPUKernelNHWCSmall(
...
         // Warp-accumulate the pixels of the same depth from the accumulator.
+        int lane_id;
+        asm volatile ("mov.u32 %0, %laneid;" : "=r"(lane_id));
+        int sub_warp = lane_id / kAccumPixels;
+        int zeros = sub_warp * kAccumPixels;
+        unsigned mask = ((1U << kAccumPixels) - 1) << zeros;
         for (int delta = kAccumPixels / 2; delta > 0; delta /= 2) {
-          val += CudaShuffleDown(val, delta);
+          val += CudaShuffleXor(mask, val, delta);
         }
```
Comment (jlebar) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: This whole chunk of code is now a lot more complicated than it was, and is repeated elsewhere.  Can we abstract it into a function?

File: tensorflow/core/util/cuda_kernel_helper.h [hunk 4]
```diff
@@ -26,6 +26,16 @@ limitations under the License.
...
 
+#if defined(__CUDACC_VER_MAJOR__) && __CUDACC_VER_MAJOR__ < 9
+__device__ inline void __syncwarp(unsigned mask=0xFFFFFFFF) {}
+
+#define __ballot_sync(mask, predicate)              __ballot(predicate)
+#define __shfl_sync(mask, val, srcLane, width)      __shfl(val, srcLane, width)
+#define __shfl_down_sync(mask, val, delta, width)   __shfl_down(val, delta, width)
+#define __shfl_up_sync(mask, val, delta, width)     __shfl_up(val, delta, width)
+#define __shfl_xor_sync(mask, val, laneMask, width) __shfl_xor(val, laneMask, width)
+#endif
```
Comment (jlebar) [inline tensorflow/core/util/cuda_kernel_helper.h]: Please add comments explaining what's going on here, and in particular what "sync" means.

File: tensorflow/core/util/cuda_kernel_helper.h [hunk 5]
```diff
@@ -26,6 +26,16 @@ limitations under the License.
 #include "tensorflow/core/platform/stream_executor.h"
 #include "tensorflow/core/platform/types.h"
 
+#if defined(__CUDACC_VER_MAJOR__) && __CUDACC_VER_MAJOR__ < 9
```
Comment (jlebar) [inline tensorflow/core/util/cuda_kernel_helper.h]: TensorFlow defines a CUDA_VERSION macro; please use that so this code will be portable to other compilers.

**Review by jlebar — DISMISSED** [review]
Thank you for doing this; it will be a big help.

What tools did you use to figure out what needed to be changed and to verify that the code now works?

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc (hunk 1 above)
Comment (zheng-xq) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: +1 about abstracting away in a function

Comment (sjperkins) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: `cub::LaneId()` does this and I believe cub is a dependency now.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc (hunk 3 above)
Comment (sjperkins) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: Available in cub I believe

File: tensorflow/core/util/cuda_kernel_helper.h (hunk 5 above)
Comment (nluehr) [reply tensorflow/core/util/cuda_kernel_helper.h]: TF_CUDA_VERSION is a string containing a decimal value (e.g. "8.0") rather than an INT. This makes it unwieldy for inequality comparisons.

Comment (jlebar) [reply tensorflow/core/util/cuda_kernel_helper.h]: What about the CUDA_VERSION macro?  https://github.com/tensorflow/tensorflow/search?utf8=%E2%9C%93&q=cuda_version&type=

Comment (nluehr) [reply tensorflow/core/util/cuda_kernel_helper.h]: Perhaps I'm being dense, but I don't see where that macro is defined, and it doesn't seem to be available in CUDA compilation units.

Comment (jlebar) [reply tensorflow/core/util/cuda_kernel_helper.h]: CUDA_VERSION is defined by nvidia's headers, in cuda.h.

(FYI I'm going to be OOO for a week; hopefully you can get assistance from the TF team in the meantime.)

**Comment by tfboyd** [comment]
Asking team about CLA for associate corporation.  Should not take long to sort out.  Thank you for starting the review.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc (hunk 2 above)
Comment (nluehr) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: cuda_kernel_helper.h now defines CUDA_WARP_ALL for use in the warp-function also defined there.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc (hunk 1 above)
Comment (nluehr) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: I wasn't initially sure of the ordering of treads in thread_idx, but on closer inspection, thread_idx % 32 would give the lane correctly. However, after abstracting the warp reduction into a separate function, it is preferable to not pass a separate thread index / lane id anyway.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc (hunk 3 above)
Comment (nluehr) [reply tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: Done.

File: tensorflow/core/util/cuda_kernel_helper.h (hunk 4 above)
Comment (nluehr) [reply tensorflow/core/util/cuda_kernel_helper.h]: Done

**Comment by jthestness** [comment]
Nice work! Thanks for these changes. I have these cuda-9 and cudnn-v7 (#12503) patches built and working.

Notes for others trying this out:
1) This patch alone either requires prior versions of cuDNN (up to v6), or cuDNN must be disabled (either option does not support new mixed-precision fp16 NVIDIA Volta V100). Another patch is required to also support cuDNN v7 (and Volta). @nluehr has posted a patch for cuDNN 7 support here: #12503

2) Eigen must also be updated to work with cuda-9 per this thread: #12489. For now, @nluehr has a modified Eigen archive available with these changes to `tensorflow/workspace.bzl`:
```
diff --git a/tensorflow/workspace.bzl b/tensorflow/workspace.bzl
index e5c2ab5..1b3314a 100644
--- a/tensorflow/workspace.bzl
+++ b/tensorflow/workspace.bzl
@@ -169,12 +169,14 @@ def tf_workspace(path_prefix="", tf_repo_name=""):
   native.new_http_archive(
       name = "eigen_archive",
       urls = [
-          "http://mirror.bazel.build/bitbucket.org/eigen/eigen/get/f3a22f35b044.tar.gz",
-          "https://bitbucket.org/eigen/eigen/get/f3a22f35b044.tar.gz",
+          "http://bazel-mirror.storage.googleapis.com/bitbucket.org/ncluehr/eigen/get/6cc72014d888.tar.gz",
+          "https://bitbucket.org/ncluehr/eigen/get/6cc72014d888.tar.gz",
       ],
-      sha256 = "ca7beac153d4059c02c8fc59816c82d54ea47fe58365e8aded4082ded0b820c4",
-      strip_prefix = "eigen-eigen-f3a22f35b044",
+      sha256 = "549fc01af565d185123016c12baf997eeb462b088c1885a4c3dfaf2a5c82d677",
+      strip_prefix = "ncluehr-eigen-6cc72014d888",
       build_file = str(Label("//third_party:eigen.BUILD")),
+
+
   )
 
   native.new_http_archive(
```

**Comment by nluehr** [comment]
I signed it!

**Comment by zheng-xq** [comment]
LGTM

**Review by zheng-xq — COMMENTED** [review]
LGTM

File: tensorflow/core/kernels/bias_op_gpu.cu.cc [hunk 6]
```diff
@@ -173,10 +173,10 @@ __global__ void BiasGradNCHW_SharedAtomics(const T* output_backprop,
   // Accumulate the results in the shared memory into the first element.
   // No syncthreads is needed since this is only in the same warp.
   int32 thread_index = threadIdx.x;
-  if (thread_index < 16) s_data[thread_index] += s_data[thread_index + 16];
-  if (thread_index < 8) s_data[thread_index] += s_data[thread_index + 8];
-  if (thread_index < 4) s_data[thread_index] += s_data[thread_index + 4];
-  if (thread_index < 2) s_data[thread_index] += s_data[thread_index + 2];
+  if (thread_index < 16) s_data[thread_index] += s_data[thread_index + 16]; __syncwarp(0xFFFF);
```
Comment (jlebar) [inline tensorflow/core/kernels/bias_op_gpu.cu.cc]: Please put the __syncwarp calls on a separate line.  As written it looks like the if statement applies to both statements, but it doesn't.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 7]
```diff
@@ -114,8 +124,10 @@ static string GetSmName(std::pair<int, int> compute_capability) {
                                                            {{5, 3}, 53},
                                                            {{6, 0}, 60},
                                                            {{6, 1}, 61},
-                                                           {{6, 2}, 62}});
-  int sm_version = 20;
+                                                           {{6, 2}, 62},
+                                                           // TODO: Change this to 70 once LLVM NVPTX supports it
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: Please keep lines <80 characters.

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc [hunk 8]
```diff
@@ -1015,6 +1016,21 @@ __global__ void __launch_bounds__(640, 2)
...
+// Device function to compute sub-warp reductions for a power-of-two group of
+// neighboring threads.
+template<int kWidth, typename T>
+__device__ __forceinline__ T WarpReduce(T val) {
+  // support only power-of-two widths.
+  assert(__popc(kWidth) == 1);
+  int sub_warp = cub::LaneId() / kWidth;
+  int zeros = sub_warp * kWidth;
+  unsigned mask = ((1U << kWidth) - 1) << zeros;
+  for (int delta = kWidth/2; delta>0; delta/=2) {
```
Comment (jlebar) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: TensorFlow style adds spaces around operators.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 9]
```diff
@@ -996,7 +996,7 @@ Status IrEmitterUnnested::EmitRowReduction(
   //   for (shuffle_distance = 16; shuffle_distance > 0; shuffle_distance /= 2)
   //     partial_result = Reducer(
   //         partial_result,
-  //         __shfl_down(partial_result, shuffle_distance));
+  //         __shfl_down_sync(0xFFFFFFFF, partial_result, shuffle_distance));
```
Comment (jlebar) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: CUDA_WARP_ALL?

File: tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc [hunk 10]
```diff
@@ -1015,6 +1016,21 @@ __global__ void __launch_bounds__(640, 2)
   }
 }
 
+// Device function to compute sub-warp reductions for a power-of-two group of
```
Comment (jlebar) [inline tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc]: "reduction" is a generic term -- what is being done here is a *sum* reduction.  I think we should change the comment, but also the name of the function.

File: tensorflow/core/util/cuda_kernel_helper.h [hunk 11]
```diff
@@ -25,6 +25,28 @@ limitations under the License.
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/platform/stream_executor.h"
 #include "tensorflow/core/platform/types.h"
+#include "cuda/include/cuda.h"
+
+#define CUDA_WARP_ALL 0xFFFFFFFF
```
Comment (jlebar) [inline tensorflow/core/util/cuda_kernel_helper.h]: Please add a comment.

Comment (jlebar) [inline tensorflow/core/util/cuda_kernel_helper.h]: Please add a comment.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 9 above)
Comment (nluehr) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Switched to CUDA_WARP_ALL. But if this is ever un-commented, a dependency on cuda_kernel_helper.h will need to be added.

**Comment by jlebar** [comment]
What tools did you use to figure out what needed to be changed and to verify that the code now works?

**Review by jlebar — APPROVED** [review]

**Comment by jlebar** [comment]
Approval, but I would really like to know what tools were used and so on so we can (ideally) add tests that this continues to be clean wrt whatever sanitizers you used

**Review by Artem-B — COMMENTED** [review]
Issue with __*_sync macros.

**Comment by pieroit** [comment]
So being this PR merged into master, should we compile by hand to use it?
How much time before it goes into the version distributed by `pip`?

Thank you :)

**Comment by byronyi** [comment]
I think tf-nightly is specifically for the impatient.

**Comment by pieroit** [comment]
thank you @byronyi, unfortunately tf-nightly is not yet (but soon will) suport GPUs. From the readme:

```We are pleased to announce that TensorFlow now offers nightly pip packages under the tf-nightly project on pypi. Simply run pip install tf-nightly in a clean environment to install the nightly tensorflow build. We currently only support CPU packages on Linux, Mac, and Windows. GPU packages on all platforms will arrive soon!```

**Comment by lefnire** [comment]
FWIW I'm seeing a tf-gpu-nightly link in https://github.com/tensorflow/tensorflow/commit/4270c4a00bb90d97418df5b0e9b6b3e148a72a1e, in case they just haven't updated the README? Worth a shot trying to install it with pip, just in case

**Comment by pieroit** [comment]
@lefnire cannot find it on pip. I guess it's a matter of days ;)

**Comment by tfboyd** [comment]
It will not be in the nightly as far as linking to CUDA 9.  TF 1.4 is next, which will **most likely** stay CUDA 8 + cuDNN 6 and then 1.5 would move to cuDNN 7 and CUDA 9 if all goes well.  cuDNN 7 + Cuda 9 gives you Volta support along with other internal changes like batch_norm supporting FP16 optimally.  There may be other benefits but Volta is the big deal.

**Comment by pieroit** [comment]
I'm gonna roll back to cuDNN 6 + CUDA 8 and wait some time. GPU computing is quite a mess :D 
Thank you guys
